src/dialects/ibmi/query-generator.js

```diff
@@ -0,0 +1,836 @@
+'use strict';
+
+const Utils = require('../../utils');
+const util = require('util');
+const _ = require('lodash');
+const AbstractQueryGenerator = require('../abstract/query-generator');
+const DataTypes = require('../../data-types');
+const Model = require('../../model');
+const SqlString = require('../../sql-string');
+
+const typeWithoutDefault = new Set(['BLOB']);
+
+class IBMiQueryGenerator extends AbstractQueryGenerator {
+
+  // Version queries
+  versionQuery() {
+    return 'SELECT CONCAT(OS_VERSION, CONCAT(\'.\', OS_RELEASE)) AS VERSION FROM SYSIBMADM.ENV_SYS_INFO';
+  }
+
+  // Schema queries
+  createSchema(schema) {
+    return `CREATE SCHEMA "${schema}"`;
+  }
+
+  dropSchema(schema) {
+    return `BEGIN IF EXISTS (SELECT * FROM SYSIBM.SQLSCHEMAS WHERE TABLE_SCHEM = ${schema ? `'${schema}'` : 'CURRENT SCHEMA'}) THEN SET TRANSACTION ISOLATION LEVEL NO COMMIT; DROP SCHEMA "${schema ? `${schema}` : 'CURRENT SCHEMA'}"; COMMIT; END IF; END`;
+  }
+
+  showSchemasQuery(options) {
+    let skippedSchemas = '';
+    if (options.skip) {
+      for (let i = 0; i < options.skip.length; i++) {
+        skippedSchemas += ` AND SCHEMA_NAME != '${options.skip[i]}'`;
+      }
+    }
+
+    return `SELECT DISTINCT SCHEMA_NAME AS "schema_name" FROM QSYS2.SYSSCHEMAAUTH WHERE GRANTEE = CURRENT USER${skippedSchemas}`;
+  }
+
+  // Table queries
+  createTableQuery(tableName, attributes, options) {
+    const primaryKeys = [];
+    const foreignKeys = {};
+    const attrStr = [];
+
+    for (const attr in attributes) {
+      if (!Object.prototype.hasOwnProperty.call(attributes, attr)) {
+        continue;
+      }
+
+      const dataType = attributes[attr];
+
+      if (dataType.includes('PRIMARY KEY')) {
+        primaryKeys.push(attr);
+        attrStr.push(`${this.quoteIdentifier(attr)} ${dataType.replace('PRIMARY KEY', '')}`);
+      } else {
+        attrStr.push(`${this.quoteIdentifier(attr)} ${dataType}`);
+      }
+    }
+
+    let attributesClause = attrStr.join(', ');
+    const pkString = primaryKeys.map(pk => this.quoteIdentifier(pk)).join(', ');
+
+    if (options.uniqueKeys) {
+      _.each(options.uniqueKeys, (columns, indexName) => {
+        // if primary keys === unique keys, then skip adding new constraint
+        const uniqueIsPrimary = columns.fields.length === primaryKeys.length && columns.fields.sort().every((value, index) => {
+          return value === primaryKeys.sort()[index];
```

Comment:
`primaryKeys.sort()` is going to be called many times


I think this piece of code would be better extracted to a `shallowEqual` function:
```typescript
columns.fields.sort();
primaryKeys.sort();

shallowEqual(columns.fields, primaryKeys);
```

---

`sort` is going to mutate columns.fields. I am generally opposed to mutating inputs as it often leads to unexpected bugs. (this is something sequelize does a lot but that I'd like to resolve)

Reply:
I think I have this fixed now. Creates a copy of `primaryKeys` to sort (one time per table), and creates a copy of `uniqueKeys.fields` to sort (on time per unique key), then does its comparison on those arrays.